packages/dds/tree/src/shared-tree/treeAlpha.ts

```diff
@@ -1051,8 +1084,106 @@ export const TreeAlpha: TreeAlpha = {
+				nodeEventName === "nodeChanged" || nodeEventName === "treeChanged",
+				0xa60 /* Invalid event for TreeNode */,
+			);
+			const nodeListener = listener as TreeChangeEvents[typeof nodeEventName];
+			return treeNodeApi.on(parent, nodeEventName, nodeListener);
+		} else {
+			// Parent is a TreeView
+			const view = parent as TreeView<ImplicitFieldSchema>;
+			if (eventName === "nodeChanged" || eventName === "treeChanged") {
+				const nodeEventName = eventName as K & keyof TreeChangeEvents;
+				const nodeListener = listener as TreeChangeEvents[typeof nodeEventName];
+				let isSubscribed = true;
+				let currentNodeUnsubscribe: (() => void) | undefined;
+
+				// Helper function to subscribe and re-subscribe to the root node.
+				const subscribeToRoot = (): void => {
+					if (!isSubscribed) {
+						return;
+					}
+
+					assert(
+						view.compatibility.canView,
+						0xa5f /* Cannot subscribe to node events on a TreeView with incompatible schema */,
+					);
+					const rootNode = view.root;
+					// rootNode may be undefined if the root is optional.
+					currentNodeUnsubscribe = isTreeNode(rootNode)
+						? treeNodeApi.on(rootNode, nodeEventName, nodeListener)
+						: undefined;
+				};
+
+				// Initial subscription
+				subscribeToRoot();
+
+				// Subscribe to rootChanged to handle cases where the root is replaced
+				const unsubscribeRootChanged = view.events.on("rootChanged", () => {
+					(listener as (...args: unknown[]) => void)();
+
+					// Unsubscribe from the old root's events
+					if (currentNodeUnsubscribe !== undefined) {
+						currentNodeUnsubscribe();
+						currentNodeUnsubscribe = undefined;
+					}
+					// Subscribe to the new root's events
+					subscribeToRoot();
+				});
+
+				// Return a combined unsubscribe function
+				return () => {
+					isSubscribed = false;
+					if (currentNodeUnsubscribe !== undefined) {
+						currentNodeUnsubscribe();
+						currentNodeUnsubscribe = undefined;
+					}
+					unsubscribeRootChanged();
+				};
+			}
+			// For view-level events (rootChanged, schemaChanged, commitApplied)
+			const viewEventName = eventName as K & keyof TreeViewEvents;
+			const viewListener = listener as TreeViewEvents[typeof viewEventName];
+			return view.events.on(viewEventName, viewListener);
+		}
+	},
 };
 
+/**
+ * This function gets the TreeView from a TreeNode.
+ * TODO: If the kernel is not hydrated, this will throw an assertion error.
+ * We may want a parent object for unhydrated nodes as well, so we will need to handle that case.
+ *
+ * @param node - The TreeNode to get the view from.
+ * @returns TreeView associated with the node.
+ */
+function getViewFromNode(node: TreeNode): TreeView<ImplicitFieldSchema> {
+	const kernel = getKernel(node);
+	assert(kernel.isHydrated(), 0xa5d /* Node must be hydrated to get its view */);
+	const view = kernel.anchorNode.anchorSet.slots.get(ViewSlot);
+	assert(view !== undefined, 0xa5e /* Expected TreeView to be present in ViewSlot */);
```

Comment:
DO not introduce multiple asserts with the same short code. Do not introduce asserts that look like they have been tagged with short codes.